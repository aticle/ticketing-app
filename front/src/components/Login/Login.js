// @flow
import React, { Component } from 'react';
import { connect, type Dispatch } from 'react-redux';
import { withRouter, RouterHistory } from 'react-router-dom';
import { TextField, Button } from '@material-ui/core';
import { loginUser } from '../../actions/authAction';
import './Login.css';

type Props = {
    errors: Array<Error>,
    loginUser: Dispatch,
    history: RouterHistory
};
type State = {
    email: string,
    password: string,
    errors: Object
};

class Login extends Component<Props, State> {
    constructor(props) {
        super(props);

        this.handleChange = this.handleChange.bind(this);
    }

    handleChange = function (name: string) {
        return (function (e: SyntheticEvent<HTMLInputElement>) {
            const value = e.currentTarget.value;

            this.setState({
                ...this.state,
                [name]: value
            });
        }).bind(this);
    }

    state = {
        email: '',
        password: '',
        errors: {}
    };


    handleSubmit = (e: Event) => {
        e.preventDefault();
        const user = {
            email: this.state.email,
            password: this.state.password
        }

        this.props.loginUser(user, this.props.history);
    }

    render() {
        const { email, password } = this.state;

        return (
            <form className="login" onSubmit={this.handleSubmit}>
                <h2>Login</h2>
                <TextField
                    label="email"
                    id="email"
                    type="email"
                    value={email}
                    required
                    onChange={this.handleChange('email')}
                ></TextField>
                <TextField
                    label="password"
                    id="password"
                    type="password"
                    value={password}
                    required
                    onChange={this.handleChange('password')}
                ></TextField>
                <Button
                    id="submit"
                    variant='contained'
                    type="submit"
                >LOGIN</Button>
            </form>
        );
    };
}

export const mapDispatchToProps = (dispatch: Dispatch) => ({
    loginUser: (user: any, history: RouterHistory) => loginUser(user, history)(dispatch)
})

export const mapStateToProps = (state: State) => ({
    errors: state.errors
})

export default connect(mapStateToProps, mapDispatchToProps)(withRouter(Login));