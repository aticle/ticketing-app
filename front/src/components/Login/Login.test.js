import React from 'react';
import { shallow } from 'enzyme';
import Login, { mapStateToProps, mapDispatchToProps } from "./Login";

const props = {
    loginUser: jest.fn(),
    history: jest.fn()
};
const login = shallow(<Login.WrappedComponent {...props} />);
const loginI = login.instance();

describe('Login', () => {
    it('renders correctly', () => {
        const login = shallow(<Login.WrappedComponent />);
        expect(login).toMatchSnapshot();
    });

    it('handleChange', () => {
        const value = 'viva@olla.com'
        const event = { currentTarget: { value } };
        const change = 'email';
        loginI.handleChange(change)(event);
        expect(loginI.state[change]).toBe(value);
    });

    it('handleSubmit', () => {
        const event = { preventDefault: jest.fn() };
        const user = {
            email: 'gedult@bringtrosen.ge',
            password: 'verstanden'
        };
        loginI.setState({
            ...loginI.state,
            ...user
        });
        loginI.handleSubmit(event);
        expect(loginI.props.loginUser.mock.calls[0]).toEqual([user, loginI.props.history]);
    });

    it('mapStateToProps', () => {
        const errors = {};
        expect(mapStateToProps({ errors })).toHaveProperty('errors', errors);
    });

    it('mapDispatchToProps', () => {
        const dispatch = jest.fn();
        const user = {
            email: 'gedult@bringtrosen.ge',
            password: 'verstanden'
        };
        mapDispatchToProps(dispatch).loginUser(user, loginI.props.history);
        expect(loginI.props.loginUser.mock.calls[0][0]).toEqual(user);
    });
});