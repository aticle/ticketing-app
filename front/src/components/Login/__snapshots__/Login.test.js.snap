// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Login />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2>
          Login
        </h2>,
        <TextField
          id="email"
          label="email"
          onChange={[Function]}
          required={true}
          select={false}
          type="email"
          value=""
          variant="standard"
        />,
        <TextField
          id="password"
          label="password"
          onChange={[Function]}
          required={true}
          select={false}
          type="password"
          value=""
          variant="standard"
        />,
        <WithStyles(Button)
          id="submit"
          type="submit"
          variant="contained"
        >
          LOGIN
        </WithStyles(Button)>,
      ],
      "className": "login",
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Login",
        },
        "ref": null,
        "rendered": "Login",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "id": "email",
          "label": "email",
          "onChange": [Function],
          "required": true,
          "select": false,
          "type": "email",
          "value": "",
          "variant": "standard",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "id": "password",
          "label": "password",
          "onChange": [Function],
          "required": true,
          "select": false,
          "type": "password",
          "value": "",
          "variant": "standard",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "LOGIN",
          "id": "submit",
          "type": "submit",
          "variant": "contained",
        },
        "ref": null,
        "rendered": "LOGIN",
        "type": [Function],
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2>
            Login
          </h2>,
          <TextField
            id="email"
            label="email"
            onChange={[Function]}
            required={true}
            select={false}
            type="email"
            value=""
            variant="standard"
          />,
          <TextField
            id="password"
            label="password"
            onChange={[Function]}
            required={true}
            select={false}
            type="password"
            value=""
            variant="standard"
          />,
          <WithStyles(Button)
            id="submit"
            type="submit"
            variant="contained"
          >
            LOGIN
          </WithStyles(Button)>,
        ],
        "className": "login",
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Login",
          },
          "ref": null,
          "rendered": "Login",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "id": "email",
            "label": "email",
            "onChange": [Function],
            "required": true,
            "select": false,
            "type": "email",
            "value": "",
            "variant": "standard",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "id": "password",
            "label": "password",
            "onChange": [Function],
            "required": true,
            "select": false,
            "type": "password",
            "value": "",
            "variant": "standard",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "LOGIN",
            "id": "submit",
            "type": "submit",
            "variant": "contained",
          },
          "ref": null,
          "rendered": "LOGIN",
          "type": [Function],
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenThreeAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
